x-pack/plugins/infra/common/plugin_config_types.ts

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+export interface InfraConfig {
+  alerting: {
+    inventory_threshold: {
+      group_by_page_size: number;
+    };
+    metric_threshold: {
+      group_by_page_size: number;
+    };
+  };
+  inventory: {
+    compositeSize: number;
+  };
+  sources?: {
+    default?: {
+      fields?: {
+        message?: string[];
+      };
+    };
+  };
+}
+
+export const publicConfigKeys = {
+  sources: true,
+} as const;
+
+export type InfraPublicConfigKey = keyof {
+  [K in keyof typeof publicConfigKeys as typeof publicConfigKeys[K] extends true ? K : never]: true;
```

Comment:
Yes, this is not great to read - sorry :grimacing: It's essentially the equivalent to

```typescript
Object.entries(publicConfigKeys)
  .filter(([key, value]) => value === true)
  .map(([key, value]) => key)
```

on the type level: We want to get the union type of all the key literals that are exposed to the browser. The source of truth for that is the `publicConfigKeys` object. We lift it into type space using `typeof`, map over its entries and conditionally remap to `never` for all non-true values. In the end we're taking the `keyof` because we just need the key names. This can then be used on the client side to ensure the compiler complains when accessing a config that is not exposed.

![image](https://user-images.githubusercontent.com/973741/160635438-a5c3e67e-3efc-4acc-8057-624c60bfa448.png)
![image](https://user-images.githubusercontent.com/973741/160636933-2f2708a4-d18a-4ed7-8da2-594d1cc751b0.png)

Does that make sense?

Reply:
Yep, makes sense. That was really helpful, thank you!